plugins/inputs/cloudwatch/cloudwatch.go

```diff
@@ -239,168 +412,208 @@ func (c *CloudWatch) updateWindow(relativeTo time.Time) error {
+				c.queryDimensions["sample_count_"+id] = dimension
+				dataQueries = append(dataQueries, &cloudwatch.MetricDataQuery{
+					Id:    aws.String("sample_count_" + id),
+					Label: aws.String(snakeCase(*metric.MetricName + "_sample_count")),
+					MetricStat: &cloudwatch.MetricStat{
+						Metric: metric,
+						Period: aws.Int64(int64(c.Period.Duration.Seconds())),
+						Stat:   aws.String(cloudwatch.StatisticSampleCount),
+					},
+				})
+			}
+		}
 	}
-	configProvider := credentialConfig.Credentials()
 
-	c.client = cloudwatch.New(configProvider)
-	return nil
-}
+	if len(dataQueries) == 0 {
+		return nil, errors.New("no metrics found to collect")
+	}
 
-/*
- * Fetch available metrics for given CloudWatch Namespace
- */
-func (c *CloudWatch) fetchNamespaceMetrics() ([]*cloudwatch.Metric, error) {
-	if c.metricCache != nil && c.metricCache.IsValid() {
-		return c.metricCache.Metrics, nil
+	c.queryCache = &queryCache{
+		queries: dataQueries,
+		built:   time.Now(),
+		ttl:     c.CacheTTL.Duration,
 	}
 
-	metrics := []*cloudwatch.Metric{}
+	return dataQueries, nil
+}
 
-	var token *string
-	for more := true; more; {
-		params := &cloudwatch.ListMetricsInput{
-			Namespace:  aws.String(c.Namespace),
-			Dimensions: []*cloudwatch.DimensionFilter{},
-			NextToken:  token,
-			MetricName: nil,
-		}
+// gatherMetrics gets metric data from Cloudwatch.
+func (c *CloudWatch) gatherMetrics(
+	params *cloudwatch.GetMetricDataInput,
+) ([]*cloudwatch.MetricDataResult, error) {
+	results := []*cloudwatch.MetricDataResult{}
 
-		resp, err := c.client.ListMetrics(params)
+	for {
+		resp, err := c.client.GetMetricData(params)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("failed to get metric data: %v", err)
 		}
 
-		metrics = append(metrics, resp.Metrics...)
-
-		token = resp.NextToken
-		more = token != nil
-	}
-
-	c.metricCache = &MetricCache{
-		Metrics: metrics,
-		Fetched: time.Now(),
-		TTL:     c.CacheTTL.Duration,
+		results = append(results, resp.MetricDataResults...)
+		if resp.NextToken == nil {
+			break
+		}
+		params.NextToken = resp.NextToken
 	}
 
-	return metrics, nil
+	return results, nil
 }
 
-/*
- * Gather given Metric and emit any error
- */
-func (c *CloudWatch) gatherMetric(
+func (c *CloudWatch) aggregateMetrics(
 	acc telegraf.Accumulator,
-	metric *cloudwatch.Metric,
+	metricDataResults []*cloudwatch.MetricDataResult,
 ) error {
-	params := c.getStatisticsInput(metric)
-	resp, err := c.client.GetMetricStatistics(params)
-	if err != nil {
-		return err
-	}
+	var (
+		grouper   = metric.NewSeriesGrouper()
+		namespace = sanitizeMeasurement(c.Namespace)
+	)
 
-	for _, point := range resp.Datapoints {
+	for _, result := range metricDataResults {
 		tags := map[string]string{
 			"region": c.Region,
-			"unit":   snakeCase(*point.Unit),
 		}
 
-		for _, d := range metric.Dimensions {
-			tags[snakeCase(*d.Name)] = *d.Value
+		if dimensions, ok := c.queryDimensions[*result.Id]; ok {
```

Comment:
Consider having queryDimensions and calling it tags, just be a map[string]string so this can be skipped.